[RDY] Build: Use GNUInstallDirs, install man pages

Pyrohh opened the pull request:
follow-up to #2605
closes #1826

For now, only install man pages matching "nvim*.1": we don't want to
install xxd.1 as it might conflict with that of a user's Vim
installation.

cc @jszakmeister @fwalch: I tested this using `DESTDIR`, then installed
it globally and everything seems to work fine, but I'm not sure if I missed
something (I'm inexperienced in CMake, as you might guess :grin:).

Pyrohh commented:
I forgot: this should be WIP because it doesn't install non-english man pages yet.

jszakmeister commented on CMakeLists.txt:
```diff
@@ -271,9 +271,20 @@ endif()
 install_helper(
   DIRECTORY runtime
-  DESTINATION share/nvim)
+  DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/nvim")
```
You don't need to add the quotes here.  CMake is smart, and even if the variables value has spaces, it will do the right thing.

Pyrohh replied:
:+1: I just read http://www.cmake.org/Wiki/CMake/Language_Syntax#Quoting, not too hard to follow.

jszakmeister commented on cmake/InstallHelpers.cmake:
```diff
@@ -107,11 +107,11 @@ function(install_helper)
   if(_install_helper_TARGETS)
     # Ensure the bin area exists with the correct permissions.
-    create_install_dir_with_perms(DESTINATION bin)
+    create_install_dir_with_perms(DESTINATION "${CMAKE_INSTALL_BINDIR}")
```
I wonder if we shouldn't be including GNUInstallDirs here somewhere too?

Pyrohh replied:
Does CMake have a notion of inheriting includes? This worked fine, so I assume something of that nature is happening.

justinmk commented:
I'm confused about the "GNU" qualifier. Will this work on BSD as well?

Pyrohh commented:
> I'm confused about the "GNU" qualifier. Will this work on BSD as well?

Heh, the name is a bit confusing, but to answer your question: yes, most likely (works fine for me on OpenBSD).

GNUInstallDirs looks to mostly consist of just directory definitions:
https://raw.githubusercontent.com/tomahawk-player/tomahawk/master/CMakeModules/GNUInstallDirs.cmake

fwalch commented:
> I forgot: this should be WIP because it doesn't install non-english man pages yet.

The non-english man pages have not been updated, so I'm not sure if we want to install them?

Pyrohh commented:
> The non-english man pages have not been updated, so I'm not sure if we want to install them?

I think you're right. I'll set this to RFC.

fwalch commented:
Tested with the Arch Linux PKGBUILD, works as expected.

`include(GNUInstallDirs)` is only required once because the variables are added to the CMake cache on the first inclusion..

What about moving `include(GNUInstallDirs)` to `InstallHelpers`?

Pyrohh commented:
> What about moving include(GNUInstallDirs) to InstallHelpers? Since this is always included when we want to install something, it would be an appropriate place IMO.

I'm not sure about that: given these are non-standard CMake variables, I think it'd be simpler for readers to keep the include in the main CMakeLists.txt. That way, they encounter the (commented) include before they encounter the non-standard variables being used.

Anyways, I'll set this to RDY, so please merge at your discretion (unless you disagree with what I said above, of course).

fwalch commented on CMakeLists.txt:
```diff
@@ -271,9 +271,20 @@ endif()
 # If it's preexisting, leave it alone.
 include(InstallHelpers)
 
+# For $CMAKE_INSTALL_{DATAROOT,MAN, ...}DIR
```
Maybe `${CMAKE_INSTALL_*_DIR}`? And maybe move it up above the `InstallHelpers` inclusion, because that already uses these variables (although currently only within a function)?

Pyrohh replied:
> Maybe ${CMAKE_INSTALL_*_DIR}?

I'll do `$CMAKE_INSTALL_*`: no need for the `{}` markup IMO.

> And maybe move it up above the InstallHelpers inclusion, because that already uses these variables (although currently only within a function)?

Sure, I'll do that. I didn't understand why you wanted to do that before, but I I understand after the above explanation.

fwalch commented:
I'm not sure if it's really simple for readers, because the variables are still used in files that don't include these variables. But it's the same with e.g. InstallHelpers, which is only included in the main `CMakeLists.txt`, so I think it's okay.

Had one more very minor comment, but feel free to merge.

Commit:
Build: Use GNUInstallDirs, install man pages

For now, only install man pages matching "nvim*.1": we don't want to
install xxd.1 as it might conflict with that of a user's Vim
installation.

closes #1826

Reviewed-by: Florian Walch <florian@fwalch.com>
Helped-by: John Szakmeister <john@szakmeister.net>

Pyrohh commented:
Merged, thanks @fwalch and @jszakmeister.

justinmk commented:
@neovim/core Has anyone been able to successfully `make install` to a `CMAKE_INSTALL_PREFIX:PATH` as a non-root user since this change?

Here's my make command:

    make CMAKE_BUILD_TYPE=RelWithDebInfo CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX:PATH=~/bin/neovim" && make install

Install fails like so:

```
-- Install configuration: "Debug"
-- Installing: /usr/local/share/man/man1/nvim.1
CMake Error at cmake_install.cmake:67 (file):
  file INSTALL cannot copy file "/Users/justin/dev/neovim/man/nvim.1" to
  "/usr/local/share/man/man1/nvim.1".
```

And after removing some lines in CMakeLists.txt:

```
-- Install configuration: "Debug"
CMake Error at cmake_install.cmake:103 (file):
  file cannot create directory: /usr/local/share/nvim/runtime/doc.  Maybe
  need administrative privileges.
```

justinmk commented:
Er, nevermind, `rm -rf build/` seems to have fixed it, I guess, though I already tried that... not sure what's up.
